junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TempDirectoryPerDeclarationTests.java

```diff
@@ -0,0 +1,981 @@
+			}));
+	}
+
+	@Test
+	@DisplayName("does not prevent constructor parameter resolution")
+	void tempDirectoryDoesNotPreventConstructorParameterResolution() {
+		executeTestsForClass(TempDirectoryDoesNotPreventConstructorParameterResolutionTestCase.class).testEvents()//
+				.assertStatistics(stats -> stats.started(1).succeeded(1));
+	}
+
+	@Test
+	@DisplayName("does not prevent user from deleting the temp dir within a test")
+	void tempDirectoryDoesNotPreventUserFromDeletingTempDir() {
+		executeTestsForClass(UserTempDirectoryDeletionDoesNotCauseFailureTestCase.class).testEvents()//
+				.assertStatistics(stats -> stats.started(1).succeeded(1));
+	}
+
+	@Test
+	@DisplayName("is capable of removal of a read-only file")
+	void nonWritableFileDoesNotCauseFailure() {
+		executeTestsForClass(NonWritableFileDoesNotCauseFailureTestCase.class).testEvents()//
+				.assertStatistics(stats -> stats.started(1).succeeded(1));
+	}
+
+	@Test
+	@DisplayName("is capable of removal of non-executable, non-writable, or non-readable directories and folders")
+	void nonMintPermissionsContentDoesNotCauseFailure() {
+		executeTestsForClass(NonMintPermissionContentInTempDirectoryDoesNotCauseFailureTestCase.class).testEvents()//
+				.assertStatistics(stats -> stats.started(13).succeeded(13));
+	}
+
+	@Test
+	@DisplayName("is capable of removal when its permissions were been changed")
+	void nonMintPermissionsDoNotCauseFailure() {
+		executeTestsForClass(NonMintTempDirectoryPermissionsDoNotCauseFailureTestCase.class).testEvents()//
+				.assertStatistics(stats -> stats.started(42).succeeded(42));
+	}
+
+	@Test
+	@DisabledOnOs(OS.WINDOWS)
+	@DisplayName("is capable of removal of a read-only file in a read-only dir")
+	void readOnlyFileInReadOnlyDirDoesNotCauseFailure() {
+		executeTestsForClass(ReadOnlyFileInReadOnlyDirDoesNotCauseFailureTestCase.class).testEvents()//
+				.assertStatistics(stats -> stats.started(1).succeeded(1));
+	}
+
+	@Test
+	@DisabledOnOs(OS.WINDOWS)
+	@DisplayName("is capable of removal of a read-only file in a dir in a read-only dir")
+	void readOnlyFileInNestedReadOnlyDirDoesNotCauseFailure() {
+		executeTestsForClass(ReadOnlyFileInDirInReadOnlyDirDoesNotCauseFailureTestCase.class).testEvents()//
+				.assertStatistics(stats -> stats.started(1).succeeded(1));
+	}
+
+	@Test
+	@DisplayName("can be used via instance field inside nested test classes")
+	void canBeUsedViaInstanceFieldInsideNestedTestClasses() {
+		executeTestsForClass(TempDirUsageInsideNestedClassesTestCase.class).testEvents()//
+				.assertStatistics(stats -> stats.started(3).succeeded(3));
+	}
+
+	@Test
+	@DisplayName("can be used via static field inside nested test classes")
+	void canBeUsedViaStaticFieldInsideNestedTestClasses() {
+		executeTestsForClass(StaticTempDirUsageInsideNestedClassTestCase.class).testEvents()//
+				.assertStatistics(stats -> stats.started(2).succeeded(2));
+	}
+
+	@Nested
+	@DisplayName("reports failure")
+	@TestMethodOrder(OrderAnnotation.class)
```

Comment:
Why the ordering here? Just to display related tests together, or is there another reason?

Reply:
@sbrannen introduced this originally in https://github.com/junit-team/junit5/commit/ab1211121927cba4837444bca35983331cd59e71.